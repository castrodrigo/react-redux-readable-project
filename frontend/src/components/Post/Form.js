import React from "react";
import PropTypes from "prop-types";
import Select from "react-select";
import styled from "styled-components";

const FormContainer = styled.form`
  max-width: 700px;
  background: #fcf7f9;
  padding: 1em;
  display: flex;
  flex-direction: column;
`;

const Input = styled.input`
  padding: 0.75em;
  width: 70%;
  border: 1px solid #012542;
  border-radius: 4px;
  box-sizing: border-box;
  &:read-only {
    background: #f6f6f6;
    color: #999999;
  }
`;
Input.displayName = "Input";

const Textarea = styled.textarea`
  padding: 0.75em;
  width: 70%;
  border: 1px solid #012542;
  border-radius: 4px;
  box-sizing: border-box;
`;
Textarea.displayName = "Textarea";

const FormSelect = styled(Select)`
  width: 70%;
  box-sizing: border-box;
  font-size: 12px;
  &:focus {
    outline: none;
  }
  & > div,
  & > div:hover {
    border-color: #012542;
  }
`;
FormSelect.displayName = "Select";

const Label = styled.label`
  background: #ffffff;
  display: flex;
  justify-content: space-between;
  margin-bottom: 20px;
  padding: 0.25em 0.5em;
  align-items: center;
`;

const Button = styled.button`
  border: 1px solid #ffffff;
  border-radius: 4px;
  color: #fff;
  background: #840032;
  padding: 10px;
  margin-top: 10px;
  &:disabled {
    background: #f0f0f0;
  }
`;
Button.displayName = "Button";

class Form extends React.Component {
  state = {
    title: "",
    author: "",
    category: null,
    body: ""
  };

  componentDidMount = () => {
    if (this.props.post) {
      const { post, categories } = this.props;
      const { title, author, category, body } = post;
      const selectedCategory = categories.find(c => c.value === category);
      this.setState({ title, author, category: selectedCategory, body });
    }
  };

  handleSubmit = event => {
    event.preventDefault();
    const {
      title,
      author,
      category: { value },
      body
    } = this.state;
    this.props.onSubmit({ title, author, category: value, body });
  };

  handleOnChange = event => {
    const {
      target: { name, value }
    } = event;
    this.setState({ [name]: value });
  };

  handleOnSelect = category => this.setState({ category });

  isButtonDisabled = () => {
    const { title, author, category, body } = this.state;
    return !(title !== "" && author !== "" && body !== "" && category !== null);
  };

  isFieldDisabled = name =>
    this.props.disabledFields && this.props.disabledFields.includes(name);

  render() {
    return (
      <FormContainer onSubmit={this.handleSubmit}>
        <Label>
          Title:
          <Input
            type="text"
            name="title"
            onChange={this.handleOnChange}
            value={this.state.title}
            readOnly={this.isFieldDisabled("title")}
          />
        </Label>
        <Label>
          Category:
          <FormSelect
            name="category"
            options={this.props.categories}
            onChange={this.handleOnSelect}
            searchable={false}
            value={this.state.category}
            isDisabled={this.isFieldDisabled("category")}
          />
        </Label>
        <Label>
          Body:
          <Textarea
            rows="10"
            name="body"
            onChange={this.handleOnChange}
            value={this.state.body}
            readOnly={this.isFieldDisabled("body")}
          />
        </Label>
        <Label>
          Author:
          <Input
            type="text"
            name="author"
            onChange={this.handleOnChange}
            value={this.state.author}
            readOnly={this.isFieldDisabled("author")}
          />
        </Label>
        <Button disabled={this.isButtonDisabled()}>Send</Button>
      </FormContainer>
    );
  }
}

Form.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  categories: PropTypes.array.isRequired,
  post: PropTypes.object
};

export default Form;
