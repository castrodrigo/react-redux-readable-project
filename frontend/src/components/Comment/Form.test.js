import React from "react";
import { shallow } from "enzyme";
import Form from "./Form";

const props = {
  onSubmit: jest.fn()
};

describe("Form", () => {
  beforeEach(() => {
    props.onSubmit = jest.fn();
  });

  describe("Render", () => {
    it("should render correctly", () => {
      const wrapper = shallow(<Form {...props} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Operations", () => {
    it("validate if button change state after filling data", () => {
      const wrapper = shallow(<Form {...props} />);

      const tBody = wrapper.find("Textarea");
      tBody.simulate("change", {
        target: { name: "body", value: "This Body" }
      });
      const iAuthor = wrapper.find("Input").last();
      iAuthor.simulate("change", {
        target: { name: "author", value: "This Author" }
      });

      const button = wrapper.find("Button").last();
      expect(button.props()).toMatchObject({ disabled: false });
    });

    it("should change state when changing input", () => {
      const wrapper = shallow(<Form {...props} />);

      const iTitle = wrapper.find("Input").first();
      iTitle.simulate("change", {
        target: { name: "title", value: "This Title" }
      });

      expect(wrapper.state("title")).toEqual("This Title");
    });

    it("should change state when changing textarea", () => {
      const wrapper = shallow(<Form {...props} />);

      const tBody = wrapper.find("Textarea");
      tBody.simulate("change", {
        target: { name: "body", value: "This Body" }
      });

      expect(wrapper.state("body")).toEqual("This Body");
    });

    it("should clear form after submit", () => {
      const wrapper = shallow(<Form onSubmit={() => Promise.resolve()} />);

      const iTitle = wrapper.find("Input").first();
      iTitle.simulate("change", {
        target: { name: "title", value: "This Title" }
      });
      const tBody = wrapper.find("Textarea");
      tBody.simulate("change", {
        target: { name: "body", value: "This Body" }
      });

      wrapper
        .find("Button")
        .last()
        .simulate("click");

      expect(iTitle.props().value).toEqual("");
      expect(tBody.props().value).toEqual("");
    });
  });
});
