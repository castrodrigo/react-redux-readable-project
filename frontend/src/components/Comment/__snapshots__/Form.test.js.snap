// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Render should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    onSubmit={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef>
          Text:
          <ForwardRef
            name="body"
            onChange={[Function]}
            rows="4"
            value=""
          />
        </ForwardRef>,
        <ForwardRef>
          Author:
          <ForwardRef
            name="author"
            onChange={[Function]}
            type="text"
            value=""
          />
        </ForwardRef>,
        <ForwardRef
          disabled={true}
        >
          Send
        </ForwardRef>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            "Text:",
            <ForwardRef
              name="body"
              onChange={[Function]}
              rows="4"
              value=""
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Text:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "name": "body",
              "onChange": [Function],
              "rows": "4",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-htpNat",
                "isStatic": true,
                "rules": Array [
                  "
  padding: 0.75em;
  width: 70%;
  border: 1px solid #012542;
  border-radius: 4px;
  box-sizing: border-box;
",
                ],
              },
              "displayName": "Textarea",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-htpNat",
              "target": "textarea",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bxivhb",
            "isStatic": true,
            "rules": Array [
              "
  background: #ffffff;
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
  padding: 0.25em 0.5em;
  align-items: center;
",
            ],
          },
          "displayName": "styled.label",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bxivhb",
          "target": "label",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            "Author:",
            <ForwardRef
              name="author"
              onChange={[Function]}
              type="text"
              value=""
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Author:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "name": "author",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bwzfXH",
                "isStatic": true,
                "rules": Array [
                  "
  padding: 0.75em;
  width: 70%;
  border: 1px solid #012542;
  border-radius: 4px;
  box-sizing: border-box;
  &:read-only {
    background: #f6f6f6;
    color: #999999;
  }
",
                ],
              },
              "displayName": "Input",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bwzfXH",
              "target": "input",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bxivhb",
            "isStatic": true,
            "rules": Array [
              "
  background: #ffffff;
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
  padding: 0.25em 0.5em;
  align-items: center;
",
            ],
          },
          "displayName": "styled.label",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bxivhb",
          "target": "label",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Send",
          "disabled": true,
        },
        "ref": null,
        "rendered": "Send",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ifAKCX",
            "isStatic": true,
            "rules": Array [
              "
  border: 1px solid #ffffff;
  border-radius: 4px;
  color: #fff;
  background: #840032;
  padding: 10px;
  margin-top: 10px;
  &:disabled {
    background: #f0f0f0;
  }
",
            ],
          },
          "displayName": "Button",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-ifAKCX",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": true,
        "rules": Array [
          "
  padding: 1em;
  display: flex;
  flex-direction: column;
",
        ],
      },
      "displayName": "styled.form",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "form",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef>
            Text:
            <ForwardRef
              name="body"
              onChange={[Function]}
              rows="4"
              value=""
            />
          </ForwardRef>,
          <ForwardRef>
            Author:
            <ForwardRef
              name="author"
              onChange={[Function]}
              type="text"
              value=""
            />
          </ForwardRef>,
          <ForwardRef
            disabled={true}
          >
            Send
          </ForwardRef>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              "Text:",
              <ForwardRef
                name="body"
                onChange={[Function]}
                rows="4"
                value=""
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Text:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "name": "body",
                "onChange": [Function],
                "rows": "4",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htpNat",
                  "isStatic": true,
                  "rules": Array [
                    "
  padding: 0.75em;
  width: 70%;
  border: 1px solid #012542;
  border-radius: 4px;
  box-sizing: border-box;
",
                  ],
                },
                "displayName": "Textarea",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htpNat",
                "target": "textarea",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": true,
              "rules": Array [
                "
  background: #ffffff;
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
  padding: 0.25em 0.5em;
  align-items: center;
",
              ],
            },
            "displayName": "styled.label",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "label",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              "Author:",
              <ForwardRef
                name="author"
                onChange={[Function]}
                type="text"
                value=""
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Author:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "name": "author",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": true,
                  "rules": Array [
                    "
  padding: 0.75em;
  width: 70%;
  border: 1px solid #012542;
  border-radius: 4px;
  box-sizing: border-box;
  &:read-only {
    background: #f6f6f6;
    color: #999999;
  }
",
                  ],
                },
                "displayName": "Input",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "input",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": true,
              "rules": Array [
                "
  background: #ffffff;
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
  padding: 0.25em 0.5em;
  align-items: center;
",
              ],
            },
            "displayName": "styled.label",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "label",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Send",
            "disabled": true,
          },
          "ref": null,
          "rendered": "Send",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": true,
              "rules": Array [
                "
  border: 1px solid #ffffff;
  border-radius: 4px;
  color: #fff;
  background: #840032;
  padding: 10px;
  margin-top: 10px;
  &:disabled {
    background: #f0f0f0;
  }
",
              ],
            },
            "displayName": "Button",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "button",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": true,
          "rules": Array [
            "
  padding: 1em;
  display: flex;
  flex-direction: column;
",
          ],
        },
        "displayName": "styled.form",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "form",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
